File: LeetCode/Top-Interview-150/Linked-List/LinkedListCycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:

        if not head or not head.next:
            return False
        
        ptr1 = head
        ptr2 = head.next

        while ptr2 and ptr2.next:
            
            if ptr1 == ptr2:

                return True
            
            ptr1 = ptr1.next
            ptr2 = ptr2.next.next

        return False
    
def createLinkedList(arr):
    if not arr:
        return None
    head = ListNode(arr[0])
    current = head
    for value in arr[1:]:
        current.next = ListNode(value)
        current = current.next
    return head

# Test the hasCycle method
solution = Solution()

# Test Case 1: No cycle
list1 = createLinkedList([1, 2, 3, 4])
print("Test Case 1 - Has Cycle:", solution.hasCycle(list1))

# Test Case 2: No cycle (Single element)
list2 = createLinkedList([1])
print("Test Case 2 - Has Cycle:", solution.hasCycle(list2))

# Test Case 3: No elements
list3 = createLinkedList([])
print("Test Case 3 - Has Cycle:", solution.hasCycle(list3))

File: LeetCode/Top-Interview-150/Linked-List/LinkedListCycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution 
{
    public boolean hasCycle(ListNode head) 
    {
        if (head == null || head.next == null) 
        {
            return false;
        }

        ListNode ptr1 = head;
        ListNode ptr2 = head.next;

        while (ptr2 != null && ptr2.next != null) 
        {
            if (ptr1 == ptr2) 
            {
                return true;
            }
            
            ptr1 = ptr1.next;
            ptr2 = ptr2.next.next;
            
        }

        return false;

    }

    public static ListNode createLinkedList(int[] arr) 
    {
        if (arr.length == 0) 
        {
                return null;
        }
        
        ListNode head = new ListNode(arr[0]);
        ListNode current = head;
        
        for (int i = 1; i < arr.length; i++) 
        {
            current.next = new ListNode(arr[i]);
            current = current.next;
        }
            return head;
    }

    public static void main(String[] args) 
    {
        Solution solution = new Solution();

        // Test Case 1: No cycle
        ListNode list1 = createLinkedList(new int[]{1, 2, 3, 4});
        System.out.println("Test Case 1 - Has Cycle: " + solution.hasCycle(list1));

        // Test Case 2: No cycle (Single element)
        ListNode list2 = createLinkedList(new int[]{1});
        System.out.println("Test Case 2 - Has Cycle: " + solution.hasCycle(list2));

        // Test Case 3: No elements
        ListNode list3 = createLinkedList(new int[]{});
        System.out.println("Test Case 3 - Has Cycle: " + solution.hasCycle(list3));
    }
}